### ISSUE TITLE
[Research] Add Dynamic Graph Attention (DGAT) for Adaptive EEG Channel Relationships

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#58]
## Paper

**Title:** DGAT: a dynamic graph attention neural network framework for EEG emotion recognition
**Authors:** Various
**Year:** 2025
**Venue:** PMC / Frontiers
**URL:** https://pmc.ncbi.nlm.nih.gov/articles/PMC12320387/

**Supporting:** EEG Emotion Recognition Using AttGraph: A Multi-Dimensional Attention-Based Dynamic Graph Convolutional Network (PMC, 2025)
**URL:** https://pmc.ncbi.nlm.nih.gov/articles/PMC12191079/

## Key Contribution

DGAT introduces *dynamic* graph attention where edge weights between EEG channels change over time and across emotional states. Unlike static graphs (issue #51 ASTAGL), the adjacency matrix is recomputed at each temporal window based on the current brain state. AttGraph adds multi-dimensional attention that evaluates sensitivity of different EEG features to emotional changes, reducing redundancy.

Key insight: the relationship between AF7-AF8 (frontal asymmetry) is emotion-dependent — the edge weight should be high during valence-relevant states but low during arousal-dominant states. Static graphs miss this.

## How It Applies to NeuralDreamWorkshop

This is a more sophisticated version of issue #51 (GNN). For the 4-channel Muse 2:
- Only 6 possible edges (4 choose 2), making dynamic attention computationally trivial
- Edge AF7-AF8 weight can increase when FAA is informative and decrease during artifacts
- Edge TP9-TP10 weight can increase during temporal asymmetry events
- The attention mechanism automatically learns *when* each channel relationship matters

This could replace the hand-coded 50% FAA + 50% alpha/beta ratio blending in emotion_classifier.py with a learned dynamic weighting.

### Files to Modify/Create
- Extends `ml/models/graph_emotion_classifier.py` from issue #51
- Add dynamic attention layer that recomputes edge weights per epoch

### Core Concept (4-Node Dynamic Attention)
```python
class DynamicGraphAttention(nn.Module):
    """
    4-node dynamic graph attention for Muse 2.
    Recomputes edge weights based on current node features.
    """
    def __init__(self, node_feat_dim=9, n_heads=2):
        super().__init__()
        self.n_heads = n_heads
        # Attention: query/key per node
        self.W_q = nn.Linear(node_feat_dim, n_heads * 8)
        self.W_k = nn.Linear(node_feat_dim, n_heads * 8)
        self.W_v = nn.Linear(node_feat_dim, n_heads * 8)
        
    def forward(self, node_features):
        """
        node_features: (batch, 4, feat_dim)
        Returns: (batch, 4, out_dim) with dynamic attention
        """
        Q = self.W_q(node_features)  # (B, 4, heads*8)
        K = self.W_k(node_features)
        V = self.W_v(node_features)
        
        # Compute attention scores (4x4 dynamic adjacency)
        attn = torch.matmul(Q, K.transpose(-1, -2)) / (8 ** 0.5)
        attn = F.softmax(attn, dim=-1)  # (B, 4, 4) — dynamic edge weights
        
        # Apply attention
        out = torch.matmul(attn, V)
        return out, attn  # attn matrix is interpretable!
```

### Interpretability Bonus
The 4x4 attention matrix at each timepoint shows which channel relationships the model uses for each prediction. This provides built-in explainability:
- High AF7-AF8 attention → model relying on frontal asymmetry (FAA)
- High TP9-AF7 attention → model relying on left-hemisphere connectivity
- Visualize attention heatmap in the dashboard

## Expected Performance Impact

- DGAT papers report 2-5% improvement over static GCN on SEED/DEAP
- For 4-channel: improvement may be smaller (fewer edges to dynamize)
- Expected: +1-3% over static GNN (issue #51), +6-18% over current LGBM
- Main value: interpretability — shows users *which brain connections* drive each emotion prediction

## Complexity: HIGH (builds on #51)

- Requires issue #51 (static GNN) to be implemented first
- Dynamic attention adds training complexity (more parameters per sample)
- With only 4 nodes, risk of overfitting the attention mechanism
- Should be implemented as a v2 upgrade to the GNN from #51
- Inference remains fast (<2ms) due to tiny graph size

## Dependencies
- Issue #51 (GNN-Based Spatial-Temporal Graph) — prerequisite
- PyTorch + torch-geometric
- Training data with temporal labels (not just per-trial)
